man/pkgadd.8.in
.TH pkgadd 8 "" "pkgutils-c @VERSION@" ""
.SH NAME
pkgadd \- install software package
.SH SYNOPSIS
\fBpkgadd [options] <file>\fP
.SH DESCRIPTION
\fBpkgadd\fP is a \fIpackage management\fP utility, which installs
a software package. A \fIpackage\fP is an archive of files (.pkg.tar.gz).
.SH OPTIONS
.TP
.B "\-u, \-\-upgrade"
Upgrade/replace package with the same name as <file>.

\fBNOTE!\fP This option is meaningless in pkgutils-c, it's used for
compatibility with C++ish pkgutils. \fBpkgadd\fP will install
new package and upgrade if it was already installed.

.TP
.B "\-o, \-\-force\-over"
Force installation, overwrite conflicting files. If the package
that is about to be installed contains files that are already
installed this option will cause all those files to be overwritten.
This option should be used with care, preferably not at all.
.TP

.B "-p, \-\-force\-perms"
Ignore permissions conflicts. By default pkgadd will fail to change
permissions on files and directories, this is done for two reasons:
you will notice when some files on your system changed permissions,
and you will in safe if you'll try to install miss-built package
(e.g. with wrong permissions on /var directory).
.TP

.B "\-f, \-\-force"
Same as if \-\-force\-over and \-\-force\-perms specified together.
.TP

.B "\-r, \-\-root <path>"
Specify alternative installation root (default is "/"). This
should \fInot\fP be used as a way to install software into
e.g. /usr/local instead of /usr. Instead this should be used
if you want to install a package on a temporary mounted partition,
which is "owned" by another system. By using this option you not only
specify where the software should be installed, but you also
specify which package database to use.
.TP
.B "\-v, \-\-version"
Print version and exit.
.TP
.B "\-h, \-\-help"
Print help and exit.
.SH CONFIGURATION
\fBpkgadd\fP is configured by the file \fI/etc/pkgadd.conf\fP. This file can contain rules, that are built out of three fragments: \fIevent\fP, \fIpattern\fP and \fIaction\fP. The event describes in what kind of situation this rule applies. Currently there are two types of events: \fBUPGRADE\fP and \fBINSTALL\fP. \fBUPGRADE\fP rules are applied when a package is installed over an existing version, and \fBINSTALL\fP rules are applied in any case. The pattern is a regular expression. The action applicable to both the \fBUPGRADE\fP and \fBINSTALL\fP event is \fBYES\fP and \fBNO\fP. More than one rule of the same
event type is allowed, in which case the first rule will have the lowest priority and the last rule
will have the highest priority. Example:

.nf
UPGRADE         "^etc/.*$"                NO
UPGRADE         "^var/log/.*$"            NO
UPGRADE         "^etc/X11/.*$"            YES
UPGRADE         "^etc/X11/XF86Config$"    NO
UPGRADE         "^etc/rc\\.local$"         NO
.fi

The above example will cause pkgadd to never upgrade anything in /etc/ or /var/log/ (subdirectories included),
except files in /etc/X11/ (subdirectories included), unless it is the file /etc/X11/XF86Config.
The default rule is to upgrade/install everything, rules in this file are exceptions to that rule.
(NOTE! A \fIpattern\fP should never contain an initial "/" since you are referring to the files in the
package, not the files on the disk. Also note that "\\" inside regular expression should be escaped, e.g.
"etc/rc\\\\.local$".)

If pkgadd finds that a specific file should not be upgraded it will install it under \fI/var/lib/pkg/rejected/\fP.
The user is then free to examine/use/remove that file manually.
.SH FILES
.TP
.B "/etc/pkgadd.conf"
Configuration file.
.SH SEE ALSO
pkgrm(8), pkginfo(8), pkgmk(8), rejmerge(8)
.SH COPYRIGHT
pkgadd (pkgutils) is Copyright (c) 2000-2005 Per Liden and is licensed through
the GNU General Public License. Read the COPYING file for the complete license.


NEWS
* 2006-11-01
Thanks to Mark Rosenstand, many bugs fixed AND pkgutils-c now autotooled!


src/pkgadd.c
//  Original pkgutils:
//  Copyright (c) 2000-2005 Per Liden
//
//  That C-rewrite:
//  Copyright (c) 2006 Anton Vorontsov <cbou@mail.ru>
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, 
//  USA.

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <pkgutils/pkgutils.h>
#include "entry.h"

static
int opt_force;

static
void print_usage(const char *argv0) {
	printf("Usage: %s [-opfrhv] <package>\n", argv0);
	puts("  -o  --force-over    ignore database and filesystem conflicts\n"
	     "  -p  --force-perms   ignore permissions conflicts\n"
	     "  -f  --force         same as -o and -p together\n"
	     "  -r  --root          specify alternate root\n"
	     "  -h  --help          display this help\n"
	     "  -v  --version       display version information");
	return;
}

static
void parse_opts(int argc, char *argv[]) {
	int c;
	struct option opts[] = {
		{"force-over" , 0, NULL, 'o'},
		{"force-perms", 0, NULL, 'p'},
		{"force"      , 0, NULL, 'f'},
		{"upgrade"    , 1, NULL, 'u'},
		{"root"       , 1, NULL, 'r'},
		{"help"       , 0, NULL, 'h'},
		{"version"    , 0, NULL, 'v'},
		{NULL         , 0, NULL, 0}
	};

	while ((c = getopt_long(argc, argv, "opfur:hv", opts, NULL)) != -1) {
		switch (c) {
			case 'f': opt_force |= PKG_ADD_FORCE_PERM;
			case 'o': opt_force |= PKG_ADD_FORCE; break;
			case 'p': opt_force |= PKG_ADD_FORCE_PERM; break;
			case 'u': break; // compatibility with C++ish pkgutils
			case 'r': opt_root = optarg; break;
			case 'h': print_usage(argv[0]); exit(0); break;
			case 'v': pkgutils_version(); exit(0); break;
			case '?': exit(1); break;
			default: break;
		}
	}

	if (optind >= argc) {
		print_usage(argv[0]);
		exit(1);
	}

	return;
}

int PKGADD_ENTRY(int argc, char *argv[]) {
	int found_conflicts;

	opt_root = "";
	parse_opts(argc, argv);

	pkg_lock_db();
	pkg_init_db();
	while (optind < argc) {
		found_conflicts = pkg_add(argv[optind], opt_force);
		if (found_conflicts & CONFLICT_PERM &&
		    !(opt_force & PKG_ADD_FORCE_PERM)) exit(1);
		else if (found_conflicts & ~CONFLICT_PERM &&
		         !(opt_force & PKG_ADD_FORCE)) exit(1);
		optind++;
	}
	pkg_free_db();
	pkg_unlock_db();

	exit(0);
	return 0;
}
